### ISSUE TITLE
Infinite recursion on integration tests when running them on iOS platform

### ISSUE THREAD

**Issue by fluiddot** [wordpress-mobile/gutenberg-mobile#5742]
**Describe the bug**
While working on https://github.com/wordpress-mobile/gutenberg-mobile/issues/5734, I spotted that some of the tests I'm writing fail without a specific reason based on the test logic. The error I'm encountering is the following one, which is triggered when initializing the editor:

```
Ran 100000 timers, and there are still more! Assuming we've hit an infinite recursion and bailing out...

      26 |
      27 | 		// Advance all timers allowing store resolvers to resolve.
    > 28 | 		act( () => jest.runAllTimers() );
         | 		                ^
      29 |
      30 | 		// The store resolvers perform several API fetches during editor
      31 | 		// initialization. The most straightforward approach to ensure all of them

      at FakeTimers.runAllTimers (node_modules/@jest/fake-timers/build/legacyFakeTimers.js:178:13)
      at runAllTimers (gutenberg/test/native/integration-test-helpers/wait-for-store-resolvers.js:28:19)
      at callback (gutenberg/node_modules/@testing-library/react-native/src/act.ts:35:24)
      at act (gutenberg/node_modules/react/cjs/react.development.js:2512:16)
      at actImplementation (gutenberg/node_modules/@testing-library/react-native/src/act.ts:34:25)
      at gutenberg/test/native/integration-test-helpers/wait-for-store-resolvers.js:28:3
      at asyncGeneratorStep (gutenberg/node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
      at _next (gutenberg/node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
      at gutenberg/node_modules/@babel/runtime/helpers/asyncToGenerator.js:32:7
      at tryCallTwo (gutenberg/node_modules/promise/lib/core.js:45:5)
      at doResolve (gutenberg/node_modules/promise/lib/core.js:200:13)
      at new Promise (gutenberg/node_modules/promise/lib/core.js:66:3)
      at gutenberg/node_modules/@babel/runtime/helpers/asyncToGenerator.js:21:12
      at fn (gutenberg/test/native/integration-test-helpers/with-fake-timers.js:21:23)
      at asyncGeneratorStep (gutenberg/node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
      at _next (gutenberg/node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
      at gutenberg/node_modules/@babel/runtime/helpers/asyncToGenerator.js:32:7
      at tryCallTwo (gutenberg/node_modules/promise/lib/core.js:45:5)
      at doResolve (gutenberg/node_modules/promise/lib/core.js:200:13)
      at new Promise (gutenberg/node_modules/promise/lib/core.js:66:3)
      at gutenberg/node_modules/@babel/runtime/helpers/asyncToGenerator.js:21:12
      at withFakeTimers (gutenberg/test/native/integration-test-helpers/with-fake-timers.js:11:26)
      at gutenberg/test/native/integration-test-helpers/wait-for-store-resolvers.js:24:9
      at asyncGeneratorStep (gutenberg/node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
      at _next (gutenberg/node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
      at gutenberg/node_modules/@babel/runtime/helpers/asyncToGenerator.js:32:7
      at tryCallTwo (gutenberg/node_modules/promise/lib/core.js:45:5)
      at doResolve (gutenberg/node_modules/promise/lib/core.js:200:13)
      at new Promise (gutenberg/node_modules/promise/lib/core.js:66:3)
      at gutenberg/node_modules/@babel/runtime/helpers/asyncToGenerator.js:21:12
      at waitForStoreResolvers (gutenberg/test/native/integration-test-helpers/wait-for-store-resolvers.js:15:33)
      at gutenberg/test/native/integration-test-helpers/initialize-editor.js:35:9
      at asyncGeneratorStep (gutenberg/node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
      at _next (gutenberg/node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
      at gutenberg/node_modules/@babel/runtime/helpers/asyncToGenerator.js:32:7
      at tryCallTwo (gutenberg/node_modules/promise/lib/core.js:45:5)
      at doResolve (gutenberg/node_modules/promise/lib/core.js:200:13)
      at new Promise (gutenberg/node_modules/promise/lib/core.js:66:3)
      at gutenberg/node_modules/@babel/runtime/helpers/asyncToGenerator.js:21:12
      at initializeEditor (gutenberg/test/native/integration-test-helpers/initialize-editor.js:27:28)
      at src/test/videopress/test.js:10:24
      at asyncGeneratorStep (node_modules/@babel/runtime/helpers/asyncToGenerator.js:3:24)
      at _next (node_modules/@babel/runtime/helpers/asyncToGenerator.js:25:9)
      at node_modules/@babel/runtime/helpers/asyncToGenerator.js:32:7
      at tryCallTwo (gutenberg/node_modules/promise/lib/core.js:45:5)
      at doResolve (gutenberg/node_modules/promise/lib/core.js:200:13)
      at new Promise (gutenberg/node_modules/promise/lib/core.js:66:3)
      at node_modules/@babel/runtime/helpers/asyncToGenerator.js:21:12
```

**To Reproduce**
1. Apply the following patch with a test example:
```patch
diff --git forkSrcPrefix/src/test/test.js forkDstPrefix/src/test/test.js
new file mode 100644
index 0000000000000000000000000000000000000000..7621b671b21ed0e5cbfb88c3157098d7e2508397
--- /dev/null
+++ forkDstPrefix/src/test/test.js
@@ -0,0 +1,17 @@
+/**
+ * External dependencies
+ */
+import { getBlock, initializeEditor, setupCoreBlocks } from 'test/helpers';
+
+setupCoreBlocks();
+
+describe( 'Test', () => {
+	test.each( [ 1, 2, 3 ] )( 'Test %i', async () => {
+		const screen = await initializeEditor( {
+			initialHtml:
+				'<!-- wp:paragraph --><p>Test</p><!-- /wp:paragraph -->',
+		} );
+		const block = await getBlock( screen, 'Paragraph' );
+		expect( block ).toBeVisible();
+	} );
+} );
```
2. Run the command: `TEST_RN_PLATFORM=ios npm run test:update test.js`
3. Observe that the above error is printed when running the tests.

**Expected behavior**
Initializing the editor on integration tests shouldn't make tests to fail.

**Screenshots**
N/A

**Smartphone (please complete the following information):**
 - Device: N/A
 - OS: N/A
 - Version N/A

**Additional context**
N/A

**Comment by fluiddot** [comment]
So far I narrowed down the cause to the iOS version of `KeyboardAwareFlatList` component, specifically [this part that throttles a function](https://github.com/WordPress/gutenberg/blob/282da7ca4c74725193c2c362925c9c8b5946b993/packages/components/src/mobile/keyboard-aware-flat-list/index.ios.js#L78C30-L92). I'll keep investigating and sharing any findings 👀 .

**Comment by dcalhoun** [comment]
I [encountered this timeout](https://github.com/WordPress/gutenberg/pull/50293#discussion_r1187346152) as well and applied a workaround. Thanks for investigating this. The root cause you identified seems very plausible. If we address the origin, it likely makes sense to revert my change.

**Comment by fluiddot** [comment]
> I [encountered this timeout](https://github.com/WordPress/gutenberg/pull/50293#discussion_r1187346152) as well and applied a workaround. Thanks for investigating this. The root cause you identified seems very plausible. If we address the origin, it likely makes sense to revert my change.

Ah, very interesting and perfect timing 🙂, thanks @dcalhoun for sharing it 🙇 !

I've just applied [the workaround you shared](https://github.com/WordPress/gutenberg/pull/50293#discussion_r1187346152) and also solves this issue 🎊 . Since I managed to find the component where it's failing, I'll keep investigating in case this leads to uncovering an issue that affects the editor.

**Comment by fluiddot** [comment]
Seems the issue is produced by the `debounce` function and the use of a recursive timer for invoking the debounced function ([reference](https://github.com/WordPress/gutenberg/blob/d770f9738a9f1bfad1fda8c2e9c4db8dc53bb04a/packages/compose/src/utils/debounce/index.ts#L192-L200)). I tested locally and confirmed that when the editor is initialized on integration tests, calling [this throttled function](https://github.com/WordPress/gutenberg/blob/282da7ca4c74725193c2c362925c9c8b5946b993/packages/components/src/mobile/keyboard-aware-flat-list/index.ios.js#L78C30-L92) leads to creating tons of timers and hitting the recursion limit. 

Following [the Jest documentation](https://jestjs.io/docs/timer-mocks#run-pending-timers), seems the best way to handle this case is by using [the solution you provided with `runOnlyPendingTimers`](https://github.com/WordPress/gutenberg/commit/34d6d78ddb17889fb69c3438eab8b425b834847c). That said, I'll close this issue in favor of that approach. Thanks for the help @dcalhoun  🙇 !

**Comment by fluiddot** [comment]
I bumped into this in a different test and after investigating a bit further I managed to find out the culprit ([reference](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5778#discussion_r1198872045)). I reviewed the `debounce` function and the fake timers logic and realized that `Date.now` is returning the real-time, even when using fake timers. `debounce` relies on this to calculate the expiration of the timers, hence despite running/advancing the timers, the expiration is never met leading to infinite recursion.

I worked on a solution to this in [this PR](https://github.com/WordPress/gutenberg/pull/50784). The idea is basically mock `Date.now` while using fake timers, and return the faked time provided by Jest.
